keps/sig-windows/4802-windows-node-shutdown/README.md

```diff
@@ -0,0 +1,539 @@
+the service control handler, the above "pre-shutdown handler" is part of this service control handler.
+
+At this "InitService" stage, the kubelet has not read the kubelet configs/flags yet, so the "nodeshutdown_manager_windows", 
+has not been instantiated, we will then set the "preshutdownhander" to be NIL, so will ignore the preshutdown event if any. 
+This should be safe as the kubelet itself has not been started, let alone the pods it manages, so there is no 
+need to do anything for shutting down at this stage.
+
+When kubelet proceeds with starting up and initialize the "nodeshutdown_manager_window", it will set the service control ehandler’s 
+pre-shutdownhandler to the "nodeshutdown_manager_window" and is now ready to monitor and process the preshutdown event
+
+We will also add function of "UpdatePreShutdownInfo" and "QueryPreShutdownInfo" to Windows service package, these functions will be 
+used by the "nodeshutdown_manager" to query and update the preshtudown timeout.
+
+If we will take the 'pre-shutdown timeout value as pariale with the 'InhibitDelayMaxSec` in linux,
+then the other parts of implementation will be pretty much similar with what linux has [implemented](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown#implementation)
+
+The poc of the implementation can be found [here](https://github.com/zylxjtu/kubernetes/commit/854ea4bde88c0905241b43f5f80d470967bb909f#diff-8494875e9a1e884afd36f32fa90dceb2b827616f1e981ecb83a12901262214c7)
+
+### Notes/Constraints/Caveats (Optional)
+
+n/a
+
+### Risks and Mitigations
+
+Please refer to the the [part](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown#notesconstraintscaveats-optional) in KEP-2000
+
+## Design Details
+
+Use the same Kubelet Config as in linux.
+```
+type KubeletConfiguration struct {
+    ...
+    ShutdownGracePeriod metav1.Duration
+    ShutdownGracePeriodCriticalPods metav1.Duration
+}
+```
+
+Communication with service control manager will make use of golang.org/x/sys/windows/svc packag,
+which is already included in [vendor](https://github.com/kubernetes/kubernetes/tree/release-1.19/vendor/golang.org/x/sys/windows)
+
+
+Termination of pods will make use of the existing
+[killPod](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/kubelet/pod_workers.go#L292) function
+from the `kubelet` package and specify the appropriate `gracePeriodOverride` as
+necessary.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+
+##### Unit tests
+
+It does not seem possible to unit test the Windows logic of setting service handler and update registry keys
```

Comment:
This needs some clarification. Are we saying that we can't simulate a shutdown using registry?

Reply:
I think this is saying that we cannot simulate setting the registry keys that control service shutdown behavior (specifically WaitToKillServiceTimeout`
From https://learn.microsoft.com/en-us/windows/win32/services/service-control-handler-function 

+1 for some clarification in the KEP

Reply:
Add more clarification